docs/architecture/adr-059-evidence-composition-and-lifecycle.md

```diff
@@ -0,0 +1,304 @@
+
+A second problem brought up in a prior [issue](https://github.com/tendermint/tendermint/issues/4593) is the notion of evidence censorship. Consider the following scenario: I have greater than 1/3 validator power. I perform an attack either on a light client of on the main network that was detected. Whenever someone proposes a block that contains evidence I abstain from voting meaning that it never is committed. When it is my turn to propose, I propose a block that does not include the evidence, everyone else votes and that block gets committed. Eventually the chain reaches a height and time where the evidence has expired and any attempt to punish my attack has been thwarted.
+
+This would most likely need an RFC (in order to explore possible solutions). As far as I can see there is a 0% approach a 50% approach and 100% approach in terms of the level of intervention.
+
+The 0% approach would just be simply to leave this as is.
+
+The 50% approach would be that validators would consider a proposal invalid if it didn't contain at least one of the evidence that that validator had and thus wouldn't vote on it. This solution is far less intrusive to the current consensus algorithm but it also doesn't quite solve the problem of censorship, rather this protocol will keep the chain halted until off-chain governance can intervene.
+
+The 100% approach is a lot more intricate and would be to find a protocol where upon receiving a valid proposal with evidence, current validators that had been shown to have misbehaved would be barred from voting such that the total voting power would be reduced by the voting power of the malicious validators and this would allow for 2/3 consensus to be reached regardless of what the malicious validators did.
+
+
+## Proposal
+
+A Hybrid design
+
+Evidence has the following simple interface:
+
+```go
+type Evidence interface {  //proposed
+  Height() int64                                     // height of the equivocation
+  Bytes() []byte                                     // bytes which comprise the evidence
+  Hash() []byte                                      // hash of the evidence
+  ValidateBasic() error
+  String() string
+}
+```
+
+We have two recognized evidence that complies with it
+
+```go
+type LightClientAttackEvidence struct {
+  ConflictingBlock *LightBlock
+  CommonHeight int64
+}
+```
+where the `Hash()` is the hash of the header and commonHeight
+
+```go
+type DuplicateVoteEvidence {
+  VoteA *Vote
+  VoteB *Vote
+}
+```
+where the `Hash()` is the hash of the two votes
+
+The first is generated in the light client, the second in consensus. Both are sent to the evidence pool through `AddEvidence(ev Evidence) error` where it first runs `Has(ev Evidence)` to check if it has already received it (by comparing hashes) and then  `Verify(ev Evidence) error` before adding it to the database
+
+
+`Verify()` does the following:
+
+- Use the hash to see if we already have this evidence in our committed database.
```

Comment:
Can you explain more about the "committed database"? When does evidence enter that database?

For more context on how I'm thinking about this, here are my notes on the first part of the light client evidence lifecycle:

1. Light client notices that something is amiss
2. Light client constructs a `LightClientAttackEvidence` object 
3. Light client passes this object to a full node--typically the full node that is serving as its "witness"
4. The full node verifies that this `LightClientAttackEvidence` is valid by calling `Verify`
5. If valid, the full node gossips the `LightClientAttackEvidence` to other full nodes. 
6. Once a proposing node has the `LightClientAttackEvidence` on hand, it will include it in the next block it proposes

Where in this process does the evidence enter a database?

Reply:
Between step 4 and 5. We persist data in case something happens to the evidence pool

Reply:
OK. So from this and the description above, along with a peek at the actual code, is it fair to say that the evidence pool is almost like a reactor itself, in that it has both a bunch of business logic as well as ways to persist data (through databases it holds onto)? 

And the evidence reactor deals with evidence that comes in BOTH from the light client and from other full nodes (like when the evidence is being gossiped)?

Reply:
I just had this discussion with Marko about this. To me the reactor is this little socket thing on the side that interfaces with the p2p world but I think to most people the reactor encompasses the pool or whatever logic the rest of the module performs. So perhaps I should call everything the evidence reactor

Reply:
"Reactor," yet another ambiguous and overloaded term in our codebase...

Reply:
Perhaps we need a `GLOSSARY.md`?